cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala

```diff
@@ -59,27 +61,54 @@ class DrsLocalizerMain(drsUrl: String,
     }
   }
 
-  def resolveAndDownloadWithRetries(retries: Int, downloaderFactory: DownloaderFactory,
+  def resolveAndDownloadWithRetries(downloadRetries: Int,
+                                    checksumRetries: Int,
+                                    downloaderFactory: DownloaderFactory,
                                     backoff: Option[CloudNioBackoff],
-                                    attempt: Int = 0): IO[DownloadResult] = {
+                                    downloadAttempt: Int = 0,
+                                    checksumAttempt: Int = 0): IO[DownloadResult] = {
 
-    def maybeResolveAndDownloadWithRetry(t: Throwable): IO[DownloadResult] = {
-      if (attempt < retries) {
+    def maybeRetryForChecksumFailure(t: Throwable): IO[DownloadResult] = {
+      if (checksumAttempt < checksumRetries) {
         backoff foreach { b => Thread.sleep(b.backoffMillis) }
-        logger.warn(s"Attempting retry $attempt of $retries retries to download $drsUrl", t)
-        resolveAndDownloadWithRetries(retries, downloaderFactory, backoff map { _.next }, attempt + 1)
+        logger.warn(s"Attempting retry $checksumAttempt of $checksumRetries checksum retries to download $drsUrl", t)
+        resolveAndDownloadWithRetries(downloadRetries, checksumRetries, downloaderFactory, backoff map { _.next }, downloadAttempt, checksumAttempt + 1)
       } else {
-        IO.raiseError(new RuntimeException(s"Exhausted $retries retries to resolve and download $drsUrl", t))
+        IO.raiseError(new RuntimeException(s"Exhausted $checksumRetries checksum retries to resolve, download and checksum $drsUrl", t))
+      }
+    }
+
+    def maybeRetryForDownloadFailure(t: Throwable): IO[DownloadResult] = {
+      t match {
+        case _: FatalRetryDisposition =>
+          IO.raiseError(t)
+        case _: TransientRetryDisposition =>
+          backoff foreach { b => Thread.sleep(b.backoffMillis) }
+          logger.warn(s"Transient failure: Attempting retry $downloadAttempt of $downloadRetries download retries to download $drsUrl", t)
+          // Do not increment the retry counter for transient errors. *Do* take the next backoff.
```

Comment:
Does this mean we could theoretically continue retrying forever, if we get a 408 or 429? Or is there a timeout elsewhere in the system?

Reply:
TOL (thinking out loud): I actually wonder if the notion of a class of error that doesn't count against the retry count is over-complicating things. Sure, someone could get unlucky through a sequence of different types of errors, but that'll probably be a rare case.

429 is an interesting one. I could imagine that we might want a completely different retry strategy for that (since it's specifically related to rate limiting)... in other words, differently configured retry intervals and allowed number of retries. Ultimately, we're trying to codify a heuristic for a sensible level of persistence for this context. I think some day we could consider refactoring the code such that the rules are isolated from the actions, and keep the rules simple in the mean time.

Reply:
Hard to judge the relative badness of the theoretical possibility we could hang forever vs. the actual cases where we'll fail a download due to transient errors we could have waited out, without knowing how often the latter happens. Since (I think?) this is happening on the VM and not in the main Cromwell process I'm less worried about it, but I also agree with Brian that keeping things simple until we have a "real" strategy we fully believe in would be good. There's also the option of a third, higher attempt limit for transient errors if we want to easily keep behavior similar to what's here while preventing infinite retries, though that's kind of icky. I assume an overall download timeout is out of scope for this particular PR.

I slightly lean toward either pulling out the special transient error handling or adding a third retry counter 🙄 . I don't feel like I have a good enough understanding of user expectations (how annoying is it to manually restart a failed workflow vs have one hang for a long time?) and how often we see these transient errors to have a preference between those. That said, I'm OK approving this as-is if you guys are.

Reply:
@jgainerdewar yes this code runs on the VM and not in the main Cromwell process. Someday there will (may) be analogous code that will run in the Cromwell process per BT-389, but this is not that. 🙂